### PR TITLE
[1/N][rollout] feat: Support rollout with dual logprobs return for QwenImage

### PR DESCRIPTION
**Pull request by chenyingshu**
### What does this PR do?

This is the first part for #336. 

It supports Rollout phase for:
- Additional llm token generation, returning token-level log-probs, and response ids & texts.
- Support two separate reward computations in rollout agent loop.
 
Main changes:
- New agent loop worker `CompositeAgentLoopWorker` in `verl_omni/agent_loop/composite_agent_loop.py`
- New rollout pipeline `QwenImagePipelineWithDualLogProb` in `verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py`.
- A UT script `tests/agent_loop/test_composite_agent_loop.py`.

See **Design & Code Changes** for changes.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

Run the test script:
```bash
bash tests/agent_loop/test_composite_agent_loop.py
```
It validates the returns of the new agent loop worker (`CompositeAgentLoopWorker`) and new single-turn agent loop (`CompositeSingleTurnAgentLoop`), which contain new fields: `rollout_llm_log_probs`, `llm_response_ids`, `text_encoder_responses`, and optional `llm_rm_scores` (when async reward is enabled).

### API and Usage Example

See the example usage in `tests/agent_loop/test_composite_agent_loop.py`. 
The key settings are as follows:
```python
# Apply new agent loop worker
AgentLoopManager.agent_loop_workers_class = ray.remote(CompositeAgentLoopWorker)

# Main configs to use new pipeline:
config.actor_rollout_ref.model.algorithm = "dual_grpo" # apply new rollout pipeline
config.actor_rollout_ref.rollout.agent.default_agent_loop = "composite_single_turn_agent" # run new single-turn rollout 
config.actor_rollout_ref.rollout.llm_calculate_log_probs = True  # return LLM token-level log-probabilities
config.actor_rollout_ref.rollout.max_new_tokens = 256  # Optional. For LLM token generation.
config.actor_rollout_ref.rollout.temperature = 0.8  # Optional. For LLM token generation.
config.actor_rollout_ref.rollout.top_k = 5  # Optional. For LLM token generation.
config.actor_rollout_ref.rollout.top_p = 0.9  # Optional. For LLM token generation.

# Optinoal. Apply reward loops to compute separate rewards for LLM and DiT.
# TWO reward loop worker handles should be set.
agent_loop_manager = AgentLoopManager.create(
    config=init_config,
    llm_client=llm_server_manager.get_client(), # Currently only support one server for one composite model
    reward_loop_worker_handles=[FirstRewardLoopWorkerHandleForDiT, SecondRewardLoopWorkerHandleForLLM],
)
```

### Design & Code Changes

Code Changes:
| function | file | new module |
| --- | --- | --- |
| Agent loop worker with LLM's extra returns | (New) `verl_omni/agent_loop/composite_agent_loop.py` | `CompositeAgentLoopWorker` | 
| New rollout pipeline | (New) `verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py` | `QwenImagePipelineWithDualLogProb` |
| Registration | `verl_omni/agent_loop/__init__.py`, `verl_omni/pipelines/__init__.py`, `verl_omni/pipelines/qwen_image_dual_grpo/__init__.py` | - |
| Configuration | `verl_omni/trainer/config/diffusion/rollout/diffusion_rollout.yaml`, `verl_omni/workers/config/diffusion/rollout.py`| - |
| Agent loop UT | `tests/agent_loop/test_composite_agent_loop.py` | - |

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by zhtmike** [comment]
ping @HaFred for this

**Comment by zhtmike** [comment]
Looks good for me

File: verl_omni/workers/rollout/replica.py [hunk 1]
```diff
@@ -24,6 +24,8 @@ class DiffusionOutput(BaseModel):
     """generated image tensor (CHW format) / video tensor (TCHW format)"""
     log_probs: Optional[Any] = None
     """logprobs of generated image/video"""
+    llm_log_probs: Optional[Any] = None
```
Comment (zhtmike) [inline verl_omni/workers/rollout/replica.py]: how about change to `ar_log_probs`

File: verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    Extends :class:`QwenImagePipelineWithDualLogProb` by autoregressively sampling
+    tokens from the Qwen2.5-VL text encoder before image diffusion.
+    The pipeline returns:
+
+    * DiT trajectory log-probabilities (``all_log_probs``) from the SDE window.
+    * Text-encoder token log-probabilities (``llm_all_log_probs``) for the generated text.
+    * The decoded text (``text_encoder_responses``).
+    """
+
+    def _get_qwen_text_response(
```
Comment (knlnguyen1802) [inline verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py:72]: Maybe can mark this as TODO to see if this can be fixed on vllm-omni side

Comment (chenyingshu) [reply verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py:72]: It may not be a good idea to integrate this function into QwenImage inference in vllm-omni. 
In QwenImage, during inference, the text encoder (Qwen2.5-VL) **forwards only once** to embed input prompts (ref: [pipeline_qwen_image.py](https://github.com/vllm-project/vllm-omni/blob/main/vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py#L483-L487)),
while here it **autoregressively forwards** to sample a sequence of tokens via transformers [generate()](https://huggingface.co/docs/transformers/v5.14.0/en/main_classes/text_generation#transformers.GenerationMixin.generate) method.

File: verl_omni/agent_loop/single_turn_agent_loop.py [hunk 3]
```diff
@@ -123,3 +124,71 @@ async def run(self, sampling_params: dict[str, Any], **kwargs) -> DiffusionAgent
             extra_fields=output.extra_fields,
         )
         return output
+
+
+@register("composite_single_turn_agent")
+class CompositeSingleTurnAgentLoop(DiffusionSingleTurnAgentLoop):
```
Comment (HaFred) [inline verl_omni/agent_loop/single_turn_agent_loop.py]: Do we really need a subclass here? Is it possible to refactor `DiffusionSingleTurnAgentLoop` to accept something like `llm_response_logprobs=getattr(output, "llm_log_probs", None),` ? Since quite a lot of duplication in `DiffusionSingleTurnAgentLoop` here

File: verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py (hunk 2 above)
Comment (knlnguyen1802) [reply verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py:72]: Ok thanks for explanation

File: verl_omni/agent_loop/single_turn_agent_loop.py (hunk 3 above)
Comment (chenyingshu) [reply verl_omni/agent_loop/single_turn_agent_loop.py]: reuse `DiffusionSingleTurnAgentLoop` now

File: verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+            top_k=top_k,
+        )
+        llm_response_ids, llm_all_log_probs, text_encoder_responses = self.generate_text_encoder_response(
+            prompt_ids=prompt_token_ids,
+            attention_mask=prompt_mask,
+            num_responses_per_prompt=num_images_per_prompt,  # reused for num responses per prompt in LLM generation
+            return_logprobs=llm_logprobs,
+            llm_kwargs=llm_kwargs,
+        )
+
```
[the remark this answers is not in the document]
Comment (chenyingshu) [reply verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py:415]: added `repetition_penalty`

File: verl_omni/agent_loop/composite_agent_loop.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+        # composite sampling configs for LLM and DiT
+        sampling_params = {
+            **_config_to_sampling_dict(config.pipeline),
+            **_config_to_sampling_dict(config.algo),
+            "logprobs": config.calculate_log_probs,
+            "temperature": config.temperature,
+            "top_p": config.top_p,
+            "top_k": config.top_k,
+            "repetition_penalty": 1.0,
```
[the remark this answers is not in the document]
Comment (chenyingshu) [reply verl_omni/agent_loop/composite_agent_loop.py]: done

**Comment by chenyingshu** [comment]
Ready for review again. @zhtmike @HaFred @knlnguyen1802

**Review by knlnguyen1802 — APPROVED** [review]
LGTM now, let's see if other has any comment

File: tests/agent_loop/test_composite_agent_loop.py
Comment (zhtmike) [inline tests/agent_loop/test_composite_agent_loop.py:1]: this should be add in gpu ci or add _prefix to enable cpu ci

File: verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+        repetition_penalty = coalesce_not_none(sampling_params.extra_args.get("repetition_penalty", None), 1.0)
+
+        llm_kwargs = dict(
+            max_new_tokens=max_new_tokens,
+            temperature=temperature,
+            top_p=top_p,
+            top_k=top_k,
+            repetition_penalty=repetition_penalty,
+        )
+        llm_response_ids, llm_all_log_probs, text_encoder_responses = self.generate_text_encoder_response(
```
[the remark this answers is not in the document]
Comment (chenyingshu) [reply verl_omni/pipelines/qwen_image_dual_grpo/vllm_omni_rollout_adapter.py:408]: The text encoder is not yet fine-tuned to enable "thinking and rewriting" capability; now we use the original prompt for simplicity. This PR is to ensure the new rollout pipeline can correctly output llm token-level log-probs.

File: tests/agent_loop/test_composite_agent_loop.py
Comment (chenyingshu) [reply tests/agent_loop/test_composite_agent_loop.py:1]: done

File: verl_omni/trainer/config/diffusion/rollout/diffusion_rollout.yaml [hunk 7]
```diff
@@ -254,6 +254,24 @@ pipeline:
...
+temperature: 1.0
+
+# Optional. LLM generation use
+top_k: 0
+
+# Optional. LLM generation use
+top_p: 1.0
+
+# Optional. LLM generation use
+repetition_penalty: 1.0
```
Comment (zhtmike) [inline verl_omni/trainer/config/diffusion/rollout/diffusion_rollout.yaml:270]: consider to refactor this in next PR, add new sub config for AR + Diffusion Log P only. @chenyingshu

File: verl_omni/workers/config/diffusion/rollout.py [hunk 8]
```diff
@@ -93,6 +98,13 @@ class DiffusionRolloutConfig(BaseConfig):
     n_gpus_per_node: int = 8
     n: int = 1
 
+    # for llm part when needed
+    temperature: float = 1.0
+    top_k: int = 0
+    top_p: float = 1.0
+    repetition_penalty: float = 1.0
+    max_new_tokens: int = 256
```
Comment (zhtmike) [inline verl_omni/workers/config/diffusion/rollout.py:106]: as above

**Review by zhtmike — APPROVED** [review]
